flash/partition_info/partition_info.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) 2023 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <stdio.h>
+#include <pico/stdlib.h>
+#include <pico/bootrom.h>
+#include "boot/picobin.h"
+
+
+#define PART_LOC_FIRST(x) ( ((x) & PICOBIN_PARTITION_LOCATION_FIRST_SECTOR_BITS) >> PICOBIN_PARTITION_LOCATION_FIRST_SECTOR_LSB )
+#define PART_LOC_LAST(x)  ( ((x) & PICOBIN_PARTITION_LOCATION_LAST_SECTOR_BITS)  >> PICOBIN_PARTITION_LOCATION_LAST_SECTOR_LSB )
+
+#define PARTITION_LOCATION_AND_FLAGS_SIZE  2
+#define PARTITION_ID_SIZE                  2
+#define PARTITION_NAME_MAX                 127
+#define PARTITION_TABLE_FIXED_INFO_SIZE    (4 + PARTITION_TABLE_MAX_PARTITIONS * (PARTITION_LOCATION_AND_FLAGS_SIZE + PARTITION_ID_SIZE))
+
+/*
+ * Stores partition table information and data read status
+ */
+typedef struct {
+    uint32_t table[PARTITION_TABLE_FIXED_INFO_SIZE];
+    uint32_t fields;
+    int partition_count;
+    uint32_t unpartitioned_space_first_sector;
+    uint32_t unpartitioned_space_last_sector;
+    uint32_t flags_and_permissions;
+    int current_partition;
+    size_t idx;
+    int status;
+} pico_partition_table_t;
+
+/*
+ * Stores information on each partition
+ */
+typedef struct {
+    uint32_t first_sector;
+    uint32_t last_sector;
+    uint32_t flags_and_permissions;
+    uint64_t partition_id;
+    char name[PARTITION_NAME_MAX + 1];  // name length is indicated by 7 bits
+} pico_partition_t;
+
+
+/*
+ * Read the partition table information.
+ *
+ * See the RP2350 datasheet 5.1.2, 5.4.8.16 for flags and structures that can be specified.
+ */
+int pico_partitions_open(pico_partition_table_t *pt) {
+    // Reads a fixed size fields
+    int rc = rom_get_partition_table_info(pt->table, sizeof(pt->table),
+                                          (PT_INFO_PT_INFO |
+                                           PT_INFO_PARTITION_LOCATION_AND_FLAGS |
+                                           PT_INFO_PARTITION_ID));
+    if (rc < 0) {
+        pt->status = rc;
+        return rc;
+    }
+
+    size_t pos = 0;
+    pt->fields = pt->table[pos++];
+    pt->partition_count = pt->table[pos++] & 0x000000FF;
+    uint32_t location = pt->table[pos++];
+    pt->unpartitioned_space_first_sector = PART_LOC_FIRST(location);
+    pt->unpartitioned_space_last_sector = PART_LOC_LAST(location);
```

Comment:
I dunno if it's worth printing out these values or not in your `main{}` ? :man_shrugging: